src/Microsoft.ML.TimeSeries/STL/InnerStl.cs

```diff
@@ -0,0 +1,322 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.TimeSeries
+{
+    internal class InnerStl
+    {
+        private readonly bool _isTemporal;
+        private double[] _seasonalComponent;
+        private double[] _trendComponent;
+        private double[] _residual;
+
+        private int _length;
+
+        // arrays for intermediate results
+        private List<double>[] _cycleSubSeries;
+        private List<double>[] _smoothedSubseries;
+
+        private double[] _s;
+        private double[] _t;
+        private double[] _detrendedY;
+        private double[] _c;
+        private double[] _deseasonSeries;
+
+        /// <summary>
+        /// the smoothing parameter for the seasonal component.
+        /// should be odd, and at least 7.
+        /// </summary>
+        private const int Ns = 9;
+
+        /// <summary>
+        /// the number of passes through the inner loop. /ref this value is set to 2, which works for many cases
+        /// </summary>
+        private const int Ni = 2;
+
+        /// <summary>
+        /// the number of robustness iterations of the outer loop
+        /// </summary>
+        private const int No = 10;
```

Comment:
Just a headsup that this member `No` is currently unused. I guess it doesn't harm to keep it here, but I wonder if perhaps it should actually be used somewhere in your logic?

Reply:
This parameter refers  to the number of outer loops. It is not used in current implementation as we simplified the logic and removed the outer loop. I intend to keep it here as these six parameters are listed in the paper to be the standard parameters of STL algorithm.

Reply:
I see. Can you please add a little comment stating this on the comment for the member `No`? simply saying that it's not used in the current implementation, but it's listed as one the 6 parameters of the paper. Thanks!

---
In reply to: [441256379](https://github.com/dotnet/machinelearning/pull/5202#discussion_r441256379) [](ancestors = 441256379)

Reply:
Comments added.

---
In reply to: [441627619](https://github.com/dotnet/machinelearning/pull/5202#discussion_r441627619) [](ancestors = 441627619,441256379)